### PR TITLE
LAION-400M processing

### PR DESCRIPTION
**Pull request by knighton**
## Issue #, if available:
CO-1415

### CODE REVIEW

File: streaming/text/convert/laion/laion400m/download_data.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#!/bin/sh
+
+wandb login
```
Comment (karan6181) [inline streaming/text/convert/laion/laion400m/download_data.sh:3]: is this mandatory? what if user doesn't have a wandb account? can we just comment this line and add a comment saying `uncomment this line to see the metric in wandb` ?

**Review by karan6181 — COMMENTED** [review]
Can you please add the basic unit test for each data type class such as `UInt8`, `UInt16`, and so on by just validating the `size`, `dtype`,  `encode()` method, and `decode()` method ? Thanks!

File: streaming/text/convert/laion/laion400m/download_data.sh (hunk 1 above)
Comment (knighton) [reply streaming/text/convert/laion/laion400m/download_data.sh]: Updated

File: streaming/text/convert/laion/laion400m/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+## Instructions:
+
+### 1. Setup.
+
+```
+pip3 install img2dataset==1.41.0
+
+apt install bwm-ng htop iotop
```
Comment (karan6181) [inline streaming/text/convert/laion/laion400m/README.md:8]: looks like this is only supported for ubuntu ? Maybe worth adding an instruction ?

File: streaming/text/convert/laion/laion400m/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+```
+
```
Comment (karan6181) [inline streaming/text/convert/laion/laion400m/README.md:10]: Can you please also add a step between point 1 and 2 on how to clone the streaming repo?

File: streaming/text/convert/laion/laion400m/convert_and_upload.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        local (str): Local directory containing shards.
+
+    Returns:
+        Iterator[int]: Each downloaded shard ID.
+    """
+    basenames = set(os.listdir(local))
+    count = len(list(filter(lambda s: s.endswith('_stats.json'), basenames)))
+    for idx in range(count):
+        stats_filename = os.path.join(local, f'{idx:05}_stats.json')
+        assert os.path.exists(stats_filename)
```
Comment (karan6181) [inline streaming/text/convert/laion/laion400m/convert_and_upload.py:65]: Please avoid `assert` statement in the src code. Better way to raise an Exception. Thanks!

File: streaming/text/convert/laion/laion400m/convert_and_upload.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+
+def upload(local: str, remote: str) -> None:
+    """Upload a shard to remote storage.
+
+    Args:
+        local (str): Path on local filesystem.
+        remote (str): Path on remote filesystem.
+    """
+    assert ' ' not in local
```
Comment (karan6181) [inline streaming/text/convert/laion/laion400m/convert_and_upload.py:179]: Please avoid all the assert statements. Thanks!

**Review by karan6181 — COMMENTED** [review]
Overall looks good. Just few minor comments. Thanks!

File: streaming/text/convert/laion/laion400m/README.md (hunk 2 above)
Comment (knighton) [reply streaming/text/convert/laion/laion400m/README.md:8]: done

File: streaming/text/convert/laion/laion400m/README.md (hunk 3 above)
Comment (knighton) [reply streaming/text/convert/laion/laion400m/README.md:10]: done

File: streaming/text/convert/laion/laion400m/convert_and_upload.py (hunk 5 above)
Comment (knighton) [reply streaming/text/convert/laion/laion400m/convert_and_upload.py:179]: done

File: streaming/text/convert/laion/laion400m/convert_and_upload.py (hunk 4 above)
Comment (knighton) [reply streaming/text/convert/laion/laion400m/convert_and_upload.py:65]: done

**Review by karan6181 — APPROVED** [review]
LGTM

**Pull request merged**